## This is the Liquid Cell Atlas pipeline from raw reads to methylation data, including extensive QC
# 
# Author: Nick Semenkovich <semenko@alum.mit.edu> https://nick.semenkovich.com
# Source: https://github.com/semenko/liquid-cell-atlas
#
# Note: There are modular ways to approach things with snakefiles — here, I explicitly prioritized
# simplicity and having one large file that runs the same analyses on all inputs. For example, there's
# no modular .yaml config.
#
# If you're looking for different approaches, you can explore:
# Snakepipes WGBS pipeline: https://snakepipes.readthedocs.io/en/latest/content/workflows/WGBS.html
# Snakemake Wrappers: https://github.com/snakemake/snakemake-wrappers/

import pathlib
import glob
import re

import pandas as pd
from snakemake.remote.HTTP import RemoteProvider as HTTPRemoteProvider

# TODO: Unify these / symlink from the data path?
CONFIG_FILE_PATH = "/home/nsemenkovich/liquid-cell-atlas/data/internal"
DATA_PATH = "/logo2/lca_data"

# Loop over the experiment definition files (e.g. melanoma.csv, crc.csv)
data_config_files = glob.glob(f"{CONFIG_FILE_PATH}/melanoma.csv")


# Store a big table of all our experiment .csvs
all_experiments_df = pd.DataFrame()

for filename in data_config_files:
    experiment = pathlib.Path(filename).stem

    # Load the .csv of our sample data
    experiment_df = pd.read_table(filename, delimiter=",").set_index("sample_id", drop=False)

    # Shove in the "experiment" as a column
    experiment_df['experiment'] = experiment

    # Append to one mega dataframe of all experiments
    all_experiments_df = pd.concat([all_experiments_df, experiment_df])


# This is a list of dicts, where each entry is a single sample like:
# {'experiment': 'melanoma', 'sample_id': 'Melanoma_B08_YURILES-14-3250',
#   'path_R1': 'YURILES-14-3250_R1.fastq.gz', 'path_R2': 'YURILES-14-3250_R2.fastq.gz',
#   'method': 'emseq', 'md5sum_R1': '4be250532…', 'md5sum_R2': '783e91bd2…'}
ALL_SAMPLES_LIST = []

# Select just the columns we want.
# orient=records returns a list of dicts
all_experiments_data = all_experiments_df[['experiment', 'sample_id', 'path_R1', 'path_R2', 'method', 'md5sum_R1', 'md5sum_R2']].to_dict(orient="records")

# For building a sample list for Snakemake
ALL_SAMPLES_LIST.extend(all_experiments_data)


##### FOR DEVELOPMENT
# Only one:
ALL_SAMPLES_LIST = [ALL_SAMPLES_LIST[0]]
print(ALL_SAMPLES_LIST)

###############
# Snakemake is a little funny.
# The first function here just defines what *output* we want when the whole pipeline is done.
###############
rule all:
    input:
        # For now, flagstat is our last step
        [f"{DATA_PATH}/{sample['experiment']}/{sample['sample_id']}/samtools/flagstat.tsv" for sample in ALL_SAMPLES_LIST]

###############
### Download & Build References
# These only run once.
###############

### Build GRCh38 reference
# Download our standard reference genome
# NOTE: This does not include patches (e.g. p14), even though the URL suggests otherwise.
# This is: GRCh38, with decoys and hs38d1, but no ALT and no patch. For a better understanding, read:
# https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GRCh38_major_release_seqs_for_alignment_pipelines/README_analysis_sets.txt

HTTPS = HTTPRemoteProvider()
rule get_reference_genome:
    input:
        # Not really ftp -- it's HTTPS/TLS
        reference_genome = HTTPS.remote("ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GRCh38_major_release_seqs_for_alignment_pipelines/GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.gz", keep_local=True)
    output:
        reference_genome = "{DATA_PATH}/reference/GRCh38.fna"
    run:
        shell("gunzip -c {input.reference_genome} > {output.reference_genome}")

# Apply both the NCBI U2AF1 fix and the Encode DAC Exclusion List
# U2AF1 masking file details: https://genomeref.blogspot.com/2021/07/one-of-these-things-doest-belong.html
# Encode DAC Exclusion List details: https://www.encodeproject.org/annotations/ENCSR636HFF/
rule mask_reference_fasta:
    conda: "env.yaml"
    input:
        u1af1_exclusion = HTTPS.remote("ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GRCh38_major_release_seqs_for_alignment_pipelines/GCA_000001405.15_GRCh38_GRC_exclusions.bed", keep_local=True),
        dac_exclusion = HTTPS.remote("encode-public.s3.amazonaws.com/2020/05/05/bc5dcc02-eafb-4471-aba0-4ebc7ee8c3e6/ENCFF356LFX.bed.gz", keep_local=True),
        reference_genome = "{DATA_PATH}/reference/GRCh38.fna"  # Or: rules.get_reference_genome.output
    output:
        masked_reference_genome = "{DATA_PATH}/reference/GRCh38-DAC-U2AF1.fna"
    log:
        maskfasta = "{DATA_PATH}/reference/maskfasta.log.txt"
    shell:
        """
        gunzip -c {input.dac_exclusion} > {DATA_PATH}/reference/DAC-Exclusion-List-ENCFF356LFX.bed
        cp {input.u1af1_exclusion} {DATA_PATH}/reference/U2AF1-GRCh38-Exclusion.bed
        # maskfasta doesn't work on stdin/stdout, or with gzipped files :/
        # Mask for DAC
        bedtools maskfasta -fi {input.reference_genome} -fo {DATA_PATH}/reference/GRCh38-DAC.fna -bed {DATA_PATH}/reference/DAC-Exclusion-List-ENCFF356LFX.bed >>{log.maskfasta} 2>&1
        # Mask for U2AF1
        bedtools maskfasta -fi {DATA_PATH}/reference/GRCh38-DAC.fna -fo {output.masked_reference_genome} -bed {DATA_PATH}/reference/U2AF1-GRCh38-Exclusion.bed >>{log.maskfasta} 2>&1
        """

### bwa meth index
# This should run once to generate indices for bwa mem (through bwa-meth)
rule bwa_meth_index:
    conda: "env.yaml"
    input:
        masked_reference_genome = rules.mask_reference_fasta.output
    output:
        "{DATA_PATH}/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t",
        "{DATA_PATH}/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.amb",
        "{DATA_PATH}/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.ann",
        "{DATA_PATH}/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.bwt",
        "{DATA_PATH}/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.pac",
        "{DATA_PATH}/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.sa",
    log:
        index = "{DATA_PATH}/reference/bwameth_index.log.txt"
    shell:
        "bwameth.py index {input.masked_reference_genome} >{log.index} 2>&1"

###############


###############
### Core Pipeline Steps
###############

### md5
# Validate checksums
rule md5sum:
    input:
        r1_file = lambda wildcards: "{DATA_PATH}/{experiment}/{sample}/raw/" + all_experiments_df.loc[wildcards.sample, 'path_R1'],
        r2_file = lambda wildcards: "{DATA_PATH}/{experiment}/{sample}/raw/" + all_experiments_df.loc[wildcards.sample, 'path_R2'],
    output:
        checksums = "{DATA_PATH}/{experiment}/{sample}/raw/known.md5",
        validated = "{DATA_PATH}/{experiment}/{sample}/raw/validated.md5"
    params:
        r1_checksum = lambda wildcards: all_experiments_df.loc[wildcards.sample, 'md5sum_R1'],
        r2_checksum = lambda wildcards: all_experiments_df.loc[wildcards.sample, 'md5sum_R2']
    shell:
        """
        echo -e "{params.r1_checksum}\t{input.r1_file}\n{params.r2_checksum}\t{input.r2_file}" > {output.checksums}
        md5sum --check {output.checksums} > {output.validated}
        """


### fastp
# We adapter trimming and poly-g filtering, but leave quality filtering / polyN filtering for later.
# NOTE: We also do read trimming here, 10-15 BP depending on EM- vs BS-seq
rule fastp:
    conda: "env.yaml"
    input:
        md5_sums = rules.md5sum.output.validated,
        r1_file = rules.md5sum.input.r1_file,
        r2_file = rules.md5sum.input.r2_file,
    output:
        fastp_pipe_output = pipe("{DATA_PATH}/{experiment}/{sample}/fastp/interleaved.fa"),
        fastp_failed = "{DATA_PATH}/{experiment}/{sample}/fastp/failed.fa",  # Length < 10
        fastp_json = "{DATA_PATH}/{experiment}/{sample}/fastp/report.json",
        fastp_html = "{DATA_PATH}/{experiment}/{sample}/fastp/report.html"
    log:
        fastp_err = "{DATA_PATH}/{experiment}/{sample}/fastp/fastp.log.txt"
    threads: 4
    params:
        # Inspired to do poly-g trimming by the NEB EM-seq pipeline:
        # https://github.com/nebiolabs/EM-seq/blob/master/em-seq.nf
        # TODO: make this selective for NextSeq-only runs?
        # e.g.:
        #    inst_name=$(zcat -f '!{fq_set.insert_read1}' | head -n 1 | cut -f 1 -d ':' | sed 's/^@//')
        #    fastq_barcode=$(zcat -f '!{fq_set.insert_read1}' | head -n 1 | sed -r 's/.*://')
        #    if [[ "${inst_name:0:2}" == 'A0' ]] || [[ "${inst_name:0:2}" == 'NS' ]] || \
        #       [[ "${inst_name:0:2}" == 'NB' ]] || [[ "${inst_name:0:2}" == 'VH' ]] ; then
        trim_poly_g_flag = "--trim_poly_g",

        # Trim specific to BS vs EM-seq input type
        trim_r1_5prime = "10",
        trim_r1_3prime = "10",
        trim_r2_5prime = "10",
        trim_r2_3prime = "10",

        minimum_length = 10,

        # TruSeq adapters (we could autodetect, but not needed currently).
        # TODO: Add adapters explicitly?         --adapter_sequence "{params.adapter_r1}" --adapter_sequence_r2 "{params.adapter_r2}" 
        # adapter_r1 = "AGATCGGAAGAGCACACGTCTGAACTCCAGTCA",
        # adapter_r2 = "AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT",
        # TODO: Add overrepresentation analysis?
    shell:
        """
        fastp --in1 {input.r1_file} --in2 {input.r2_file} \
        --trim_front1 {params.trim_r1_5prime} --trim_tail1 {params.trim_r1_3prime} \
        --trim_front2 {params.trim_r2_5prime} --trim_tail2 {params.trim_r2_3prime} \
        --length_required {params.minimum_length} --disable_quality_filtering {params.trim_poly_g_flag} \
        --json "{output.fastp_json}" --html "{output.fastp_html}" \
        --thread {threads} --verbose --failed_out {output.fastp_failed} --stdout \
        2>{log.fastp_err} >{output.fastp_pipe_output}
        """

### bwa_meth
# This uses bwa mem, instead of bwa-mem2
# NOTE: You could also run bwa mem2. I don't here because of a compile-time bug:
# see https://github.com/bwa-mem2/bwa-mem2/issues/207
rule bwa_meth:
    conda: "env.yaml"
    input:
        input_fasta = "{DATA_PATH}/{experiment}/{sample}/fastp/interleaved.fa",
        reference_genome = rules.mask_reference_fasta.output.masked_reference_genome,
        bwa_meth_indices = rules.bwa_meth_index.output
    output:
        pipe("{DATA_PATH}/{experiment}/{sample}/bwameth/raw.sam")
    log:
        bwameth_err = "{DATA_PATH}/{experiment}/{sample}/bwameth/log.txt"
    threads: 4
    shell:
        # TODO: FIX
        # include --read-group @RG\\tID:${fastq_barcode}\\tSM:!{fq_set.library} ?
        "bwameth.py -p -t {threads} --reference {input.reference_genome} {input.input_fasta} 2>{log.bwameth_err} >{output}"


### mark nonconverted reads
# This is a clever little program from NEB that marks non-converted reads:
# https://github.com/nebiolabs/mark-nonconverted-reads
# This sets XX:Z:UC (the X? Y? and Z? fields of SAM/BAM are user-reserved). We do *not* set the Vendor Failed bit.
# Note that  MethylDackel can independently identify poor conversion (set minConversionEfficiency).
rule mark_nonconverted:
    conda: "env.yaml"
    input:
        aligned_sam = "{DATA_PATH}/{experiment}/{sample}/bwameth/raw.sam",
        reference_genome = rules.mask_reference_fasta.output.masked_reference_genome
    output:
        pipe("{DATA_PATH}/{experiment}/{sample}/bwameth/nonconv-marked.sam")
    log:
        nonconverted = "{DATA_PATH}/{experiment}/{sample}/bwameth/nonconverted.log.txt"
    params:
        threshold = 3 # If 3 nonconverted Cs on a read, consider it nonconverted
    shell:
        # TODO: Fix this path
        "cat {input.aligned_sam} | /home/nsemenkovich/sandbox/mark-nonconverted-reads/mark-nonconverted-reads.py --c_count {params.threshold} --reference {input.reference_genome} 2>{log.nonconverted} >{output}"

### sort
# Coordinate sort our input sam, and stream as a sam (no bam yet)
rule samtools_sort:
    conda: "env.yaml"
    input:
        "{DATA_PATH}/{experiment}/{sample}/bwameth/nonconv-marked.sam"
    output:
        pipe("{DATA_PATH}/{experiment}/{sample}/samtools/sorted.sam")
    log:
        sort = "{DATA_PATH}/{experiment}/{sample}/samtools/sort.log.txt"
    shell:
        "cat {input} | samtools sort --output-fmt SAM --threads {threads} - - 2>{log.sort} >{output}"


### fixmate
# We need to ad ms and MC tags for markdup to run next
rule samtools_fixmate:
    conda: "env.yaml"
    input:
        "{DATA_PATH}/{experiment}/{sample}/samtools/sorted.sam"
    output:
        pipe("{DATA_PATH}/{experiment}/{sample}/samtools/fixmate.sam")
    log:
        fixmate = "{DATA_PATH}/{experiment}/{sample}/samtools/fixmate.log.txt"
    shell:
        # -u uncompressed output
        # -m add mate score tag
        "cat {input} | samtools fixmate -u -m --output-fmt SAM --threads {threads} - - 2>{log.fixmate} >{output}"


### markdup
# Mark duplicates from a .bam file
rule samtools_markdup:
    conda: "env.yaml"
    input:
        "{DATA_PATH}/{experiment}/{sample}/samtools/fixmate.sam"
    output:
        bam = "{DATA_PATH}/{experiment}/{sample}/aligned.bam",
        stats_file = "{DATA_PATH}/{experiment}/{sample}/samtools/markdup.txt",
    log:
        markdup = "{DATA_PATH}/{experiment}/{sample}/samtools/markdup.log.txt"
    shell:
        # TODO: Set mark supplementary alignments of duplicates as duplicates? (-S)
        # TODO: Specify index name if requred? "-##idx##indexname"
        # -f: Write stats to file
        # Output with compress level 1 (lowest)
        "cat {input} | samtools markdup -f {output.stats_file} --output-fmt BAM --output-fmt-option level=1 --write-index --threads {threads} - {output.bam} 2>{log.markdup}"

### flagstat
# Generate run statistics
rule samtools_flagstat:
    conda: "env.yaml"
    input:
        "{DATA_PATH}/{experiment}/{sample}/aligned.bam"
    output:
        "{DATA_PATH}/{experiment}/{sample}/samtools/flagstat.tsv"
    log:
        flagstat = "{DATA_PATH}/{experiment}/{sample}/samtools/flagstat.log.txt"
    shell:
        "samtools flagstat {input} 2>{log.flagstat} >{output}"


###### NOT YET IMPLEMENTED BELOW ########

### MethylDackel mbias
# Make per-sample mbias plots
rule methyldackel_mbias_plots:
    input:
    output:
    shell:
        "methyldackel mbias"

### DeepTools bias plots:
# ? Are these useful
rule deeptools_bias_plots:
    input:
    output:
    shell:
        "computeGCBias"

### wgbs_tools
# Generate pat/beta file formats
rule wgbs_tools_prep:
    input:
    output:
    shell:
        "wgbs_tools"

### multiqc
# Aggregate all our pretty data
rule multiqc:
    input:
    output:
    shell:
        "multiqc"
