## This is the Liquid Cell Atlas pipeline from raw reads to methylation data, including extensive QC
# 
# Author: Nick Semenkovich <semenko@alum.mit.edu> https://nick.semenkovich.com
# Source: https://github.com/semenko/liquid-cell-atlas
#
# Note: There are modular ways to approach things with snakefiles — here, I explicitly prioritized
# simplicity and having one large file that runs the same analyses on all inputs. For example, there's
# no modular .yaml config.
#
# If you're looking for different approaches, you can explore:
# Snakepipes WGBS pipeline: https://snakepipes.readthedocs.io/en/latest/content/workflows/WGBS.html
# Snakemake Wrappers: https://github.com/snakemake/snakemake-wrappers/

import pathlib
import glob
import re

import pandas as pd
from snakemake.remote.HTTP import RemoteProvider as HTTPRemoteProvider

# TODO: Unify these / symlink from the data path?
CONFIG_FILE_PATH = "/home/nsemenkovich/liquid-cell-atlas/data/internal"
DATA_PATH = "/logo2/lca_data"

# Loop over the experiment definition files (e.g. melanoma.csv, crc.csv)
data_config_files = glob.glob(f"{CONFIG_FILE_PATH}/melanoma.csv")

# This is a list of dicts, where each entry is a single sample like:
# {'experiment': 'melanoma', 'sample_id': 'Melanoma_B08_YURILES-14-3250',
#   'path_R1': 'YURILES-14-3250_R1.fastq.gz', 'path_R2': 'YURILES-14-3250_R2.fastq.gz',
#   'method': 'emseq', 'md5sum_R1': '4be250532…', 'md5sum_R2': '783e91bd2…'}
ALL_SAMPLES_LIST = []

for filename in data_config_files:
    experiment = pathlib.Path(filename).stem
    print(experiment)
    # Load the .csv of our sample data
    sample_df = pd.read_table(filename, delimiter=",").set_index("sample_id", drop=False)

    # Shove in the "experiment" as a column
    sample_df['experiment'] = experiment

    # Select just the columns we want.
    # orient=records returns a list of dicts
    sample_data = sample_df[['experiment', 'sample_id', 'path_R1', 'path_R2', 'method', 'md5sum_R1', 'md5sum_R2']].to_dict(orient="records")

    ALL_SAMPLES_LIST.extend(sample_data)

# Snakemake is a little funny.
# The first function here just defines what *output* we want when the whole pipeline is done.
rule all:
    input:
        # For now, flagstat is our last step
        expand("{data_path}/{sample.experiment}/{sample.sample_id}/samtools/flagstat.tsv", sample=ALL_SAMPLES_LIST, data_path=DATA_PATH)

###############
### Download & Build References
# These only run once.
###############

### Build GRCh38 reference
# Download our standard reference genome
# NOTE: This does not include patches (e.g. p14), even though the URL suggests otherwise.
# This is: GRCh38, with decoys and hs38d1, but no ALT and no patch. For a better understanding, read:
# https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GRCh38_major_release_seqs_for_alignment_pipelines/README_analysis_sets.txt

HTTPS = HTTPRemoteProvider()
rule get_reference_genome:
    input:
        # Not really ftp -- it's HTTPS/TLS
        reference_genome = HTTPS.remote("ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GRCh38_major_release_seqs_for_alignment_pipelines/GCA_000001405.15_GRCh38_no_alt_plus_hs38d1_analysis_set.fna.gz", keep_local=True)
    output:
        reference_genome = "{DATA_PATH}/reference/GRCh38.fna"
    run:
        shell("gunzip -c {input.reference_genome} > {output.reference_genome}")

# Apply both the NCBI U2AF1 fix and the Encode DAC Exclusion List
# U2AF1 masking file details: https://genomeref.blogspot.com/2021/07/one-of-these-things-doest-belong.html
# Encode DAC Exclusion List details: And also mask the DAC Exclusion List Regions
rule mask_reference_fasta:
    input:
        u1af1_exclusion = HTTPS.remote("ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GRCh38_major_release_seqs_for_alignment_pipelines/GCA_000001405.15_GRCh38_GRC_exclusions.bed", keep_local=True),
        dac_exclusion = HTTPS.remote("www.encodeproject.org/files/ENCFF356LFX/@@download/ENCFF356LFX.bed.gz", keep_local=True),
        reference_genome = "{DATA_PATH}/reference/GRCh38.fna"  # Or: rules.get_reference_genome.output
    output:
        masked_reference_genome = "{DATA_PATH}/reference/GRCh38-DAC-U2AF1.fna"
    run:
        shell("gunzip -c {input.dac_exclusion} > {DATA_PATH}/reference/DAC-Exclusion-List-ENCFF356LFX.bed")
        shell("cp {input.u1af1_exclusion} {DATA_PATH}/reference/U2AF1-GRCh38-Exclusion.bed")
        # maskfasta doesn't work on stdin/stdout, or with gzipped files :/
        # Mask for DAC
        shell("bedtools maskfasta -fi {input.reference_genome} -fo {DATA_PATH}/reference/GRCh38-DAC.fna -bed {DATA_PATH}/reference/DAC-Exclusion-List-ENCFF356LFX.bed")
        # Mask for U2AF1
        shell("bedtools maskfasta -fi {DATA_PATH}/reference/GRCh38-DAC.fna -fo {output.masked_reference_genome} -bed {DATA_PATH}/reference/U2AF1-GRCh38-Exclusion.bed")

### bwa meth index
# This should run once to generate indices for bwa mem (through bwa-meth)
rule bwa_meth_index:
    input:
        masked_reference_genome = rules.mask_reference_fasta.output
    output:
        "{DATA_PATH}/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t",
        "{DATA_PATH}/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.amb",
        "{DATA_PATH}/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.ann",
        "{DATA_PATH}/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.bwt",
        "{DATA_PATH}/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.pac",
        "{DATA_PATH}/reference/GRCh38-DAC-U2AF1.fna.bwameth.c2t.sa",
    run:
        shell("bwameth.py index {input.masked_reference_genome}")

###############


###############
### Core Pipeline Steps
###############

### md5
# Validate checksums
rule md5sum:
    input:
        r1_file = "",
        r2_file = "",
        r1_checksum = "",
        r2_checksum = "",
    output:
        "validated.md5"
    run:
        shell("checksums.md5")
        shell("md5sum --check > {output}")


### fastp
# We adapter trimming and poly-g filtering, but leave quality filtering / polyN filtering for later.
# NOTE: We also do read trimming here, 10-15 BP depending on EM- vs BS-seq
rule fastp:
    input:
        sample_pe1 = "reads/pe/{sample}.1.fastq",
        sample_pe2 = "reads/pe/{sample}.2.fastq"
    output:
        fastp_pipe_output = pipe("{sample}.fastp.interleaved"),
        fastp_failed = "report/pe/{sample}.failed.fa",  # Length < 10
        fastp_json = "report/pe/{sample}.json",
        fastp_html = "report/pe/{sample}.html"
    log:
        fastp_err = "logs/fastp/pe/{sample}.log"
    params:
        # Inspired to do poly-g trimming by the NEB EM-seq pipeline:
        # https://github.com/nebiolabs/EM-seq/blob/master/em-seq.nf
        # TODO: make this selective for NextSeq-only runs?
        # e.g.:
        #    inst_name=$(zcat -f '!{fq_set.insert_read1}' | head -n 1 | cut -f 1 -d ':' | sed 's/^@//')
        #    fastq_barcode=$(zcat -f '!{fq_set.insert_read1}' | head -n 1 | sed -r 's/.*://')
        #    if [[ "${inst_name:0:2}" == 'A0' ]] || [[ "${inst_name:0:2}" == 'NS' ]] || \
        #       [[ "${inst_name:0:2}" == 'NB' ]] || [[ "${inst_name:0:2}" == 'VH' ]] ; then
        trim_poly_g_flag = "--trim_poly_g",

        # Trim specific to BS vs EM-seq input type
        trim_r1_5prime = "10",
        trim_r1_3prime = "10",
        trim_r2_5prime = "10",
        trim_r2_3prime = "10",

        minimum_length = 10,

        # TruSeq adapters (we could autodetect, but not needed currently).
        adapter_r1 = "AGATCGGAAGAGCACACGTCTGAACTCCAGTCA",
        adapter_r2 = "AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT",
    shell:
        """
        fastp --in1 {input.sample_pe1} --in2 {input.sample_pe2} \
        --trim_front1 {params.trim_r1_5prime} --trim_tail1 {params.trim_r1_3prime} \
        --trim_front2 {params.trim_r2_5prime} --trim_tail2 {params.trim_r2_3prime} \
        --adapter_sequence {params.adapter_r1} --adapter_sequence_r2 {params.adapter_r2} \
        --length_required {params.minimum_length} --disable_quality_filtering --overrepresentation_analysis \
        {trim_polyg} \
        --json "{output.fastp_json}" --html "{output.fastp_html}" \
        --thread {snakemake.threads} \
        --stdout --failed_out {output.fastp_failed} \
        2> {log.fastp_err} > {output.fastp_pipe_output}
        """

### bwa_meth
# This uses bwa mem, instead of bwa-mem2
# NOTE: You could also run bwa mem2. I don't here because of a compile-time bug:
# see https://github.com/bwa-mem2/bwa-mem2/issues/207
rule bwa_meth:
    input:
        ".fastp.fastq",
        # Not really an input, but forces the index to build.
        bwa_meth_indices = rules.bwa_meth_index.output
    output:
        pipe(".bwameth.sam")
    shell:
        "bwameth.py -p -t !{task.cpus} --read-group @RG\\tID:${fastq_barcode}\\tSM:!{fq_set.library} --reference !{genome} /dev/stdin 2>something"


### mark nonconverted reads
# This is a clever little program from NEB that marks non-converted reads:
# https://github.com/nebiolabs/mark-nonconverted-reads
# This sets XX:Z:UC (the X? Y? and Z? fields of SAM/BAM are user-reserved). We do *not* set the Vendor Failed bit.
# Note that  MethylDackel can independently identify poor conversion (set minConversionEfficiency).
rule mark_nonconverted:
    input:
        ".bwameth.sam"
    output:
        pipe(".bwameth.nonconv.sam")
    shell:
        "mark-nonconverted-reads.py"

### sort
# Coordinate sort our input sam, and stream as a sam (no bam yet)
rule samtools_sort:
    input:
        ".bwameth.nonconv.sam"
    output:
        pipe(".sorted.sam")
    shell:
        "samtools sort --output-fmt SAM --threads {snakemake.threads} {input} {output}"


### fixmate
# We need to ad ms and MC tags for markdup to run next
rule samtools_fixmate:
    input:
        ".sorted.sam"
    output:
        pipe(".fixmate.bam")
    shell:
        # -u uncompressed output
        # -m add mate score tag
        "samtools fixmate -u -m --output-fmt SAM --threads {snakemake.threads} {input} {output}"


### markdup
# Mark duplicates from a .bam file
rule samtools_markdup:
    input:
        ".fixmate.sam"
    output:
        bam = "{sample}.bam",
        stats_file = "{sample}.txt",
    run:
        # TODO: Set mark supplementary alignments of duplicates as duplicates? (-S)
        # -f: Write stats to file
        # 
        # Output with compress level 1 (lowest)
        # "-##idx##indexname" 
        shell("samtools markdup -f {output.stats_file} --output-fmt BAM --output-fmt-option level=1 --write-index --threads {snakemake.threads} {input} {output.bam}")

### flagstat
# Generate run statistics
rule samtools_flagstat:
    input:
        "{sample}.bam"
    output:
        "{sample}.txt"
    run:
        shell("samtools flagstat {input} > {output}")





###### NOT YET IMPLEMENTED BELOW ########

### MethylDackel mbias
# Make per-sample mbias plots
rule methyldackel_mbias_plots:
    input:
    output:
    shell:
        "methyldackel mbias"

### DeepTools bias plots:
# ? Are these useful
rule deeptools_bias_plots:
    input:
    output:
    shell:
        "computeGCBias"

### wgbs_tools
# Generate pat/beta file formats
rule wgbs_tools_prep:
    input:
    output:
    shell:
        "wgbs_tools"

### multiqc
# Aggregate all our pretty data
rule multiqc:
    input:
    output:
    shell:
        "multiqc"


# Get the file sizes and add to a file
rule get_size:
    input:
        "/logo2/lca_data/melanoma/{sample}"
    output:
        "/logo2/lca_output/{sample}.txt"
    shell:
        "ls -lha {input} > {output}"
