# Sequencing Data

A collection of our sequencing data. Raw files aren't stored on Github (internally, they're symlinked / mounted from different volumes). Our published data can be downloaded via ***.

These directories are parsed by snakemake and the preprocessing/analytical pipelines.
  
| Folder | Experiment Description |
| --- | --- |
| _external | External data from Blueprint & Encode. |
| test | A single test sample for processing. |
| reference | Reference genome mapping data, generated by the Snakemake pipeline. |
| --- | --- | 
| epcam_spike | PBMCs from healthy donors, spiked with varying percentages of HT-29 DNA (an EPCAM+ CRC cell line), sequenced with EM-seq and BS-seq. |
| healthy_pbmc | PBMCs from healthy donors, sometimes with cytometry data available. |
| melanoma | Melanoma patient samples as part of a collaboration with Yale, including patients undergoing immunotherapy. |
| crc | Colorectal tumor samples, including tumor, cfDNA, and sorted populations. |

## Data Structure

Raw data files are processed and analyzed by our [Snakemake workflow](../workflow). Within each project directory, the output is (roughly) structured as:

    project/   	   	        # e.g. melanoma / crc / healthy_pbmc / epcam_spike, etc.
    │   aligned.sam             # The final alignment file (along with indices)
    ├── raw/
    │   ├── ...fastq.gz         # Raw reads
    │   └── subsampled/         # Subsampled reads for rapid analyses
    ├── bwameth/                # bwameth logs 
    ├── fastp/                  # fastp statistics & logs
    ├── fastqc/                 # fastqc graphs 
    ├── goleft/                 # goleft coverage plots
    ├── methyldackel/           # mbias plots
    ├── multiqc/                # multiqc stats
    ├── samtools/               # samtools log data
    └── wgbs_tools/             # wgbs_tools .bam/.pat files

## Sample Definition Files

Each internal experiment has an associated .csv with key experimental metadata and information. Most samples were sequenced by [MedGenome](https://research.medgenome.com/), though some were sequenced by the [McDonnell Genome Institute](https://www.genome.wustl.edu/). Of note, the "batch" in each sample is arbitrary -- and just a unique identifier for samples run on the same batch (e.g. batch B00 was not necessarily sequenced before or after batch B01). 

## WashU / Chaudhuri Lab Users

The raw data backing these files are stored in RIS in Active/lca_data (e.g. `/storage1/fs1/aadel/Active/lca_data`), which is periodically mirrored to our local compute infrastructure (currently on aclm350 in `/logo2/lca_data`).
