# Sequencing Data

A collection of our sequencing data. Raw files aren't stored on Github (internally, they're symlinked / mounted from different volumes). Our published data can be downloaded via [this Globus collection](https://app.globus.org/file-manager/collections/7e4e5850-6a7c-11ed-8fd2-e9cb7c15c7d2/overview).

These directories are parsed by snakemake and the preprocessing/analytical pipelines.
  
| Folder | Experiment Description |
| --- | --- |
| _external | External data from Blueprint & Encode. |
| test | A single test sample for processing. |
| reference | Reference genome mapping data, generated by the Snakemake pipeline. |
| --- | --- | 
| spikeCRC | PBMCs from a healthy donor (ID 70), spiked with varying percentages of CRC tumor DNA (CD45 EPCAM+ from patient 1506) sequenced with BS-seq. Questionably failed due to mixing gDNA and cfDNA? |
| spikeCRCv2 | PBMCs from a healthy donor (ID 99), spiked with varying percentages of CRC tumor DNA (CD45 EPCAM+ from patient 1484) sequenced with BS-seq. Questionably failed due to intermixing CD45+ from tumor and PBMC? |
| spikeEPCAM | PBMCs from healthy donors, spiked with varying percentages of HT-29 DNA (an EPCAM+ CRC cell line), sequenced with EM-seq and BS-seq. |
| healthy | PBMCs from healthy donors, sometimes with cytometry data available. |
| melanoma | Melanoma patient samples as part of a collaboration with Yale, including patients undergoing immunotherapy. |
| crc | Colorectal tumor samples, including tumor, cfDNA, and sorted populations. |
| --- | --- |
| *_subsampled | The Snakemake pipline generates stably subsampled raw reads (5M reads/pair, 10M total) for more rapid analysis and experimentation. |

## Data Structure

Raw data files are processed and analyzed by our [Snakemake workflow](../workflow). Within each project directory, the output is (roughly) structured as:

    project/                    # e.g. melanoma / crc / healthy, etc.
    │   SAMPLE.bam              # The final alignment file (along with indices)
    ├── fastp/                  # fastp statistics & logs
    ├── fastqc/                 # fastqc graphs 
    ├── goleft/                 # goleft coverage plots
    ├── logs/                   # runlogs from each pipeline component
    ├── methyldackel/           # mbias plots
    ├── multiqc/                # multiqc stats
    ├── raw/
    │   ├── ...fastq.gz         # Raw reads
    |   ├── ...md5.txt          # Checksums and validation
    ├── samtools/               # samtools statistics
    └── wgbs_tools/             # wgbs_tools .bam/.pat files

Note each project also has a `_subsampled` directory with identical structure, which is the result of the pipeline run on only 10M reads/sample.

## Sample Definition Files

Each internal experiment has an associated .csv with key experimental metadata and information. Most samples were sequenced by [MedGenome](https://research.medgenome.com/), though some were sequenced by the [McDonnell Genome Institute](https://www.genome.wustl.edu/). Of note, the "batch" in each sample is arbitrary -- and just a unique identifier for samples run on the same batch (e.g. batch B00 was not necessarily sequenced before or after batch B01). 

## WashU / Chaudhuri Lab Users

The raw data backing these files are stored in RIS in Active/liquid-cell-atlas (e.g. `/storage1/fs1/aadel/Active/liquid-cell-atlas `), which is periodically mirrored to our local compute infrastructure (currently in `/r5n/liquid-cell-atlas`).
